torch/csrc/autograd/record_function.h

```diff
@@ -281,36 +277,92 @@ TORCH_API void TEST_unsetGlobalSamplingProbability();
   RECORD_FUNCTION_WITH_SCOPE( \
     torch::autograd::profiler::RecordScope::USER_SCOPE, fn, {})
 
+// Notes:
+//  - for performance and interface simplicity reasons we split the set
+//    of callbacks (observers) into the two subsets - thread local and
+//    global callbacks:
+//     - thread local callbacks are added/removed only for the given thread
+//       and are stored locally for each thread and separately from the list
+//       of the global callbacks
+//     - global callbacks are stored in a single per process list and are
+//       invoked by every RecordFunction, in addition to the thread local
+//       callbacks specific to the given thread
+//     - we allow global callbacks to be sampled, by specifying a sampling
+//       probability for each global callback pair, if the start callback is
+//       not picked to run, the corresponding end callback won't be called
+//     - a typical use case for the global callbacks is passive monitoring
+//       in the background (e.g. fleet-wide monitoring), without focusing on
+//       the specific peice of code
+//     - in contrast, thread local callbacks are enabled locally, on demand,
+//       for specific piece of code (range) and are not sampled
+//     - a typical use case for thread local callbacks is profiler and
+//       code execution tracer
+//       - note, that some functionality (e.g. profiler) can automatically
```

Comment:
so the logic is that profiler specifically propagates itself, but in general we do not propagate callbacks from out thread to another, right? I'd clarify this

Reply:
yes, for now this is only some thread local callbacks, I'll add a comment